docs-internal/generic-mapping.md

```diff
@@ -0,0 +1,249 @@
+# Generic Mapping Proposal
+
+Provide an API to map an arbitrary Java type to another arbitrary Java type.
+
+## Proposal
+
+The API consist of the following classes (may be changed due to implementation details):
+- `MapperManager` - the entry point to mapping of types
+- `Mapper` - a class capable of converting one type to another
+- `MapperProvider` - SPI class to support providers loaded through Java Service loader or configured through a builder
+- `MapperException` - `RuntimeException` thrown when a mapping is missing or the mapping itself failed
+
+This API should be also added to `Config` as an additional source of mappings.
+
+TODO an alternative name would be `Converter` instead of `Mapper`
+
+### Mapper Manager
+
+Mapping provides tools to map a type to another type.
+
+The types may be
+ - a java class (`String.class`)
+ - a generic type (`io.helidon.common.GenericType` for any java type, such as `Supplier<String>`)
+ 
+Mappings can be provided either through Java Service loader, or through
+ explicitly configured providers using a builder.
+ 
+The mapping function gets either a pair of `Class` objects, or a pair of `GenericType` objects
+ that define the `SOURCE` and the `TARGET` of the mapping.
+ 
+For `Class` parameteres, lookup is done as follows:
+1. Ask each mapping provider if such a pair of classes is supported, if yes, use the first mapper
+2. Convert each class to `GenericType` and ask each mapping provider if supported, if yes, use the first mapper
+
+For `GenericType` parameters, lookup is done as follows:
+1. Ask each mapping provider if such a pair of types is supported, if yes, use the first mapper
+2. If both generic types represent a `Class`, convert each to a `Class` and 
+    ask each mapping provider if supported, if yes, use the first mapper
+    
+The results are cached (so lookup for a defined pair is done only once).
+In case no mapper is found, the result should be cached as well.
+
+The main API class is `MapperManager`.
+```java
+/**
+ * Map from source to target.
+ *
+ * @param source     object to map
+ * @param sourceType type of the source object (to locate the mapper)
+ * @param targetType type of teh target object (to locate the mapper)
+ * @return result of the mapping
+ * @throws io.helidon.common.mapping.MapperException in case the mapper was not found or failed
+ */
+<SOURCE, TARGET> TARGET map(SOURCE source, GenericType<SOURCE> sourceType, GenericType<TARGET> targetType);
+
+/**
+ * Map from source to target.
+ *
+ * @param source     object to map
+ * @param sourceType class of the source object (to locate the mapper)
+ * @param targetType class of teh target object (to locate the mapper)
+ * @return result of the mapping
+ * @throws io.helidon.common.mapping.MapperException in case the mapper was not found or failed
+ */
+<SOURCE, TARGET> TARGET map(SOURCE source, Class<SOURCE> sourceType, Class<TARGET> targetType);
```

Comment:
Can I pass in an `Integer` reference with `Number.class` as the second parameter?  If not (hazy on the generics here), then consider eliminating the `sourceType` parameter since it would then be redundant.

Reply:
No you cannot. The type mapping is exact - so if you have a mapper Number -> String and you pass an Integer, mapping would fail

Reply:
We could probably change the lookup, though it would be much more complicated, as we would have to look for a mapper of each super class of the source type. 
Assume the following example:
`CharSequence chars = mapper.map(149, CharSequence.class)`
what if there is a mapper for Number -> String - is that a suitable one?

Reply:
I just realised that this can already be implemented due to the nature of `MapperProvider`.
The method has two parameteres - the source type and the desired type.
How you implement the lookup is really up to you (or the implementor of that provider).
So if there is a  source: `Intger.class` and target: `CharSequence.class` requested, and you are capable of mapping any number into a string, you could return a mapper for this pair. This is then cached by the `MapperManager`.